### PR TITLE
Improve batch decoding

### PR DESCRIPTION
**Pull request by takaaki-hori**

### CODE REVIEW

File: espnet/bin/asr_recog.py [hunk 1]
```diff
@@ -71,6 +71,8 @@ def get_parser():
                         help='Input length ratio to obtain min output length')
     parser.add_argument('--ctc-weight', type=float, default=0.0,
                         help='CTC weight in joint decoding')
+    parser.add_argument('--ctc-window-margin', type=int, default=0,
+                        help='Use CTC window with margin parameter')
```
Comment (ShigekiKarita) [inline espnet/bin/asr_recog.py]: Can you add more description on this parameter? e.g., purpose, tradeoff, speed, (0 disables this feature)

**Comment by ShigekiKarita** [comment]
> Can you make a test for your additional changes?

Maybe the test can be just regression that detects breaking change in the future like https://github.com/espnet/espnet/blob/8cbde3843ebec4d882d5ebf35aebda7ca72b7e80/test/test_recog.py#L80

File: espnet/nets/ctc_prefix_score.py [hunk 2]
```diff
@@ -18,7 +18,18 @@ class CTCPrefixScoreTH(object):
...
+    def __init__(self, x, xlens, blank, eos, beam, scoring_ratio=1.5, margin=0):
+        """Construct CTC prefix scorer
+
+        :param x     : input label posterior sequences (B, T, O)
+        :param xlens : input length (B,)
+        :param blank : blank label id
+        :param eos   : end-of-sequence id
+        :param beam  : beam size
+        :param scoring_ratio : ratio of #scored hypos to beam size
+        :param margin: margin parameter for windowing (0 means no windowing)
```
Comment (ShigekiKarita) [inline espnet/nets/ctc_prefix_score.py]: can you specify types of arguments?

File: espnet/nets/ctc_prefix_score.py [hunk 3]
```diff
@@ -27,84 +38,152 @@ def __init__(self, x, blank, eos, beam, hlens, device_id):
...
-        for si in six.moves.range(self.n_bb):
-            log_psi[si, self.eos] = r_sum[si, self.hlens[si]]
+    def index_select_state(self, state, best_ids):
+        """Select CTC states according to best ids
 
-        return log_psi, r
+        :param state    : CTC state
+        :param best_ids : index numbers selected by beam pruning (B, W)
+        :return selected_state
+        """
```
Comment (ShigekiKarita) [inline espnet/nets/ctc_prefix_score.py]: ditto

**Comment by sw005320** [comment]
It's better to have a test of `ctc_window_margin` > 0.
Can you also report some results (RTFs and WERs?) when enabling `ctc_window_margin`?

**Comment by sw005320** [comment]
This PR is quite important and I'd like to merge it soon. If you add the test for `ctc_window_margin > 0`, it is great.

**Comment by ShigekiKarita** [comment]
LGTM except for conflicting files. It looks easy to resolve.

**Comment by sw005320** [comment]
Many thanks!
If you give us some examples and speed analysis in the other PRs later, that is great.

**Pull request merged**